From e45f457e72dea3e895a585a8dac78378e16ce37f Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Tue, 1 Nov 2016 01:43:07 +0100
Subject: [PATCH 09/22] net: dsa: add destroy callback

propagate the shutdown/destroy event to the swith driver. This allows QCA8k
to do an ordered shutdown upon reboot.

Signed-off-by: John Crispin <john@phrozen.org>
---
 include/net/dsa.h |    5 +++++
 net/dsa/dsa.c     |    4 ++++
 2 files changed, 9 insertions(+)

Index: linux-4.9.34/include/net/dsa.h
===================================================================
--- linux-4.9.34.orig/include/net/dsa.h
+++ linux-4.9.34/include/net/dsa.h
@@ -256,6 +256,11 @@ struct dsa_switch_ops {
 	u32	(*get_phy_flags)(struct dsa_switch *ds, int port);
 
 	/*
+	 * Shutdown
+	 */
+	void	(*destroy)(struct dsa_switch *ds);
+
+	/*
 	 * Access to the switch's PHY registers.
 	 */
 	int	(*phy_read)(struct dsa_switch *ds, int port, int regnum);
Index: linux-4.9.34/net/dsa/dsa.c
===================================================================
--- linux-4.9.34.orig/net/dsa/dsa.c
+++ linux-4.9.34/net/dsa/dsa.c
@@ -519,6 +519,9 @@ static void dsa_switch_destroy(struct ds
 		hwmon_device_unregister(ds->hwmon_dev);
 #endif
 
+	if (ds->ops->destroy)
+		ds->ops->destroy(ds);
+
 	/* Destroy network devices for physical switch ports. */
 	for (port = 0; port < DSA_MAX_PORTS; port++) {
 		if (!(ds->enabled_port_mask & (1 << port)))
@@ -1030,6 +1033,7 @@ static int dsa_remove(struct platform_de
 
 static void dsa_shutdown(struct platform_device *pdev)
 {
+	dsa_remove(pdev);
 }
 
 static int dsa_switch_rcv(struct sk_buff *skb, struct net_device *dev,
